File: internal/git/git.go
package git

import (
	"context"
	"errors"
	"path/filepath"
	"strings"

	"moyuan-code/internal/logging"
	"moyuan-code/internal/process"
	"moyuan-code/internal/workspace"
)

type Status struct {
	IsRepo bool     `json:"isRepo"`
	Dirty  bool     `json:"dirty"`
	Branch *string  `json:"branch"`
	Remote *string  `json:"remote"`
	Ahead  *int     `json:"ahead"`
	Behind *int     `json:"behind"`
	Files  []string `json:"files"`
}

func IsRepo(ctx context.Context, rootDir string) bool {
	res := process.RunCommand(ctx, rootDir, "git", "rev-parse", "--is-inside-work-tree")
	return res.Code == 0 && strings.TrimSpace(res.Stdout) == "true"
}

func StatusOf(ctx context.Context, rootDir string) Status {
	if !IsRepo(ctx, rootDir) {
		return Status{IsRepo: false, Dirty: false, Files: []string{}}
	}
	branch := strings.TrimSpace(process.RunCommand(ctx, rootDir, "git", "branch", "--show-current").Stdout)
	remoteRes := process.RunCommand(ctx, rootDir, "git", "remote", "get-url", "origin")
	status := process.RunCommand(ctx, rootDir, "git", "status", "--short")
	files := []string{}
	for _, line := range strings.Split(status.Stdout, "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			files = append(files, line)
		}
	}
	var branchPtr *string
	if branch != "" {
		branchPtr = &branch
	}
	var remotePtr *string
	if remoteRes.Code == 0 {
		remote := strings.TrimSpace(remoteRes.Stdout)
		if remote != "" {
			remotePtr = &remote
		}
	}
	return Status{IsRepo: true, Dirty: len(files) > 0, Branch: branchPtr, Remote: remotePtr, Files: files}
}

func Branches(ctx context.Context, rootDir string) []string {
	if !IsRepo(ctx, rootDir) {
		return []string{}
	}
	res := process.RunCommand(ctx, rootDir, "git", "branch", "--format", "%(refname:short)")
	if res.Code != 0 {
		return []string{}
	}
	branches := []string{}
	for _, line := range strings.Split(res.Stdout, "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			branches = append(branches, line)
		}
	}
	return branches
}

func Clone(ctx context.Context, url string, destDir string) error {
	res := process.RunCommand(ctx, "", "git", "clone", url, destDir)
	if res.Code != 0 {
		if strings.TrimSpace(res.Stderr) != "" {
			return errors.New(strings.TrimSpace(res.Stderr))
		}
		return errors.New("git clone failed")
	}
	return nil
}

func BindLocal(rootDir string) error {
	ws, err := workspace.Load(rootDir)
	if err != nil {
		return err
	}
	ws.Repository.Repository.Source.Type = "local_path"
	ws.Repository.Repository.Source.Provider = "local"
	ws.Repository.Repository.Source.LocalPath = rootDir
	ws.Repository.Repository.Source.URL = nil
	ws.Repository.Repository.Source.ClonePath = nil
	if err := workspace.SaveRepository(rootDir, ws.Repository); err != nil {
		return err
	}
	return logging.Log(rootDir, "git", "repository.bound.local", map[string]any{"rootDir": rootDir})
}

func BindRemote(rootDir string, url string, provider string) error {
	ws, err := workspace.Load(rootDir)
	if err != nil {
		return err
	}
	if provider == "" {
		provider = "generic_git"
	}
	ws.Repository.Repository.Source.Type = "remote_git"
	ws.Repository.Repository.Source.Provider = provider
	ws.Repository.Repository.Source.URL = &url
	ws.Repository.Repository.Source.LocalPath = ""
	ws.Repository.Repository.Source.ClonePath = nil
	if err := workspace.SaveRepository(rootDir, ws.Repository); err != nil {
		return err
	}
	return logging.Log(rootDir, "git", "repository.bound.remote", map[string]any{"url": url, "provider": provider})
}

func Sync(ctx context.Context, rootDir string) (map[string]any, error) {
	if !IsRepo(ctx, rootDir) {
		return nil, errors.New("not a git repository: " + rootDir)
	}
	fetch := process.RunCommand(ctx, rootDir, "git", "fetch", "--all", "--prune")
	if fetch.Code != 0 {
		return nil, errors.New(strings.TrimSpace(fetch.Stderr))
	}
	status := StatusOf(ctx, rootDir)
	return map[string]any{
		"branch": status.Branch,
		"remote": status.Remote,
	}, nil
}

func DefaultRemoteProjectDir(rootDir string, remote string) string {
	name := strings.TrimSuffix(filepath.Base(remote), ".git")
	if name == "" || name == "." || name == "/" {
		name = "remote-project"
	}
	return filepath.Join(rootDir, ".moyuan", "projects", name)
}

File: src/core/git.js
import path from 'node:path';
import { fileExists, readYaml, writeYaml } from '../lib/fs.js';
import { runCommand } from '../lib/process.js';
import { getWorkspacePaths } from './paths.js';
import { logEvent } from './logging.js';

export async function isGitRepo(rootDir) {
  const result = await runCommand('git', ['rev-parse', '--is-inside-work-tree'], { cwd: rootDir });
  return result.code === 0 && result.stdout.trim() === 'true';
}

export async function getGitStatus(rootDir) {
  if (!(await isGitRepo(rootDir))) {
    return {
      isRepo: false,
      dirty: false,
      branch: null,
      remote: null,
      ahead: null,
      behind: null,
      files: []
    };
  }

  const branch = await runCommand('git', ['branch', '--show-current'], { cwd: rootDir });
  const remote = await runCommand('git', ['remote', 'get-url', 'origin'], { cwd: rootDir });
  const status = await runCommand('git', ['status', '--short'], { cwd: rootDir });

  return {
    isRepo: true,
    dirty: status.stdout.trim().length > 0,
    branch: branch.stdout.trim() || null,
    remote: remote.code === 0 ? remote.stdout.trim() : null,
    ahead: null,
    behind: null,
    files: status.stdout
      .split('\n')
      .map((line) => line.trim())
      .filter(Boolean)
  };
}

export async function listBranches(rootDir) {
  if (!(await isGitRepo(rootDir))) {
    return [];
  }

  const result = await runCommand('git', ['branch', '--format', '%(refname:short)'], { cwd: rootDir });
  if (result.code !== 0) {
    return [];
  }
  return result.stdout
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean);
}

export async function currentBranch(rootDir) {
  const result = await runCommand('git', ['branch', '--show-current'], { cwd: rootDir });
  return result.code === 0 ? result.stdout.trim() || null : null;
}

export async function defaultBranch(rootDir) {
  const remote = await runCommand('git', ['symbolic-ref', 'refs/remotes/origin/HEAD'], { cwd: rootDir });
  if (remote.code === 0) {
    return path.basename(remote.stdout.trim());
  }
  return null;
}

export async function cloneRepository(url, destDir) {
  const result = await runCommand('git', ['clone', url, destDir]);
  if (result.code !== 0) {
    throw new Error(result.stderr.trim() || `git clone failed for ${url}`);
  }
  return destDir;
}

export async function bindLocalRepository(rootDir, repositoryConfig) {
  const paths = getWorkspacePaths(rootDir);
  const next = {
    ...(repositoryConfig ?? {}),
    schema_version: 1,
    repository: {
      ...(repositoryConfig?.repository ?? {}),
      source: {
        ...(repositoryConfig?.repository?.source ?? {}),
        type: 'local_path',
        provider: 'local',
        local_path: rootDir,
        url: null,
        clone_path: null
      }
    }
  };

  await writeYaml(paths.repositoryYaml, next);
  await logEvent(rootDir, 'git', 'repository.bound.local', { rootDir });
  return next;
}

export async function registerRemoteRepository(rootDir, repositoryConfig, url, provider = 'generic_git') {
  const paths = getWorkspacePaths(rootDir);
  const next = {
    ...(repositoryConfig ?? {}),
    schema_version: 1,
    repository: {
      ...(repositoryConfig?.repository ?? {}),
      source: {
        ...(repositoryConfig?.repository?.source ?? {}),
        type: 'remote_git',
        provider,
        url,
        local_path: null
      }
    }
  };

  await writeYaml(paths.repositoryYaml, next);
  await logEvent(rootDir, 'git', 'repository.bound.remote', { url, provider });
  return next;
}

export async function syncRepository(rootDir) {
  if (!(await isGitRepo(rootDir))) {
    throw new Error(`Not a git repository: ${rootDir}`);
  }

  const fetch = await runCommand('git', ['fetch', '--all', '--prune'], { cwd: rootDir });
  if (fetch.code !== 0) {
    throw new Error(fetch.stderr.trim() || 'git fetch failed');
  }

  return {
    branch: await currentBranch(rootDir),
    defaultBranch: await defaultBranch(rootDir)
  };
}

export async function createBranch(rootDir, branchName, base = null) {
  const args = ['checkout', '-b', branchName];
  if (base) {
    args.push(base);
  }
  const result = await runCommand('git', args, { cwd: rootDir });
  if (result.code !== 0) {
    throw new Error(result.stderr.trim() || `git branch create failed: ${branchName}`);
  }
  await logEvent(rootDir, 'git', 'branch.created', { branchName, base });
  return branchName;
}

export async function gitStatusText(rootDir) {
  if (!(await isGitRepo(rootDir))) {
    return 'not a git repository';
  }
  const result = await runCommand('git', ['status', '--short', '--branch'], { cwd: rootDir });
  return result.stdout.trim();
}

export async function ensureRemoteConfigured(rootDir, remote = 'origin') {
  if (!(await fileExists(rootDir))) {
    throw new Error(`Path does not exist: ${rootDir}`);
  }
  const result = await runCommand('git', ['remote', 'get-url', remote], { cwd: rootDir });
  return result.code === 0;
}
